### PR TITLE
[feature] Allow admins to expire remote public keys; refetch expired keys on demand

### PR DESCRIPTION
**Pull request by tsmethurst**
# Description

> If this is a code change, please include a summary of what you've coded, and link to the issue(s) it closes/implements.
>
> If this is a documentation change, please briefly describe what you've changed and why.

This pull request adds a new field to our account model, `public_key_expires_at`.

Using this field, it's now possible to manually expire a remote account's public key, which means that the next time your server receives a signed request from that account, it will attempt to fetch a new public key by calling the remote instance, and replace the public key in the database with the new one.

Refetching looks a bit like this in the logs:

```
timestamp="06/09/2023 17:25:36.832" func=federation.(*federator).derefPubKey level=INFO requestedUsername=tobi pubKeyID=https://queer.af/actor#main-key requestID=qsa1jtwa04001ycjm0pg msg="public key was cached, but expired at 2023-09-06 15:08:42.872193 +0000 +0000, trying dereference of new public key"
timestamp="06/09/2023 17:25:36.857" func=httpclient.(*Client).DoSigned level=INFO method=GET url=https://queer.af/actor#main-key requestID=qsa1jtwa04001ycjm0pg pubKeyID=https://goblin.technology/users/tobi/main-key msg="performing request"
timestamp="06/09/2023 17:25:36.897" func=federation.(*federator).derefPubKey level=INFO requestedUsername=tobi pubKeyID=https://queer.af/actor#main-key requestID=qsa1jtwa04001ycjm0pg msg="obtained a new public key to replace expired key, caching now; authorization for this request will be attempted with both old and new keys"
```

The PR also adds a route to the settings panel that allows admins to submit an action to mark all keys from a domain as expired, as a treat.

Screenshot of the settings panel, showing the new 'keys' expiry action.

closes https://github.com/superseriousbusiness/gotosocial/issues/2170

## Checklist

Please put an x inside each checkbox to indicate that you've read and followed it: `[ ]` -> `[x]`

If this is a documentation change, only the first checkbox must be filled (you can delete the others if you want).

- [x] I/we have read the [GoToSocial contribution guidelines](https://github.com/superseriousbusiness/gotosocial/blob/main/CONTRIBUTING.md).
- [x] I/we have discussed the proposed changes already, either in an issue on the repository, or in the Matrix chat.
- [x] I/we have performed a self-review of added code.
- [x] I/we have written code that is legible and maintainable by others.
- [x] I/we have commented the added code, particularly in hard-to-understand areas.
- [x] I/we have made any necessary changes to documentation.
- [x] I/we have added tests that cover new code.
- [x] I/we have run tests and they pass locally with the changes.
- [x] I/we have run `go fmt ./...` and `golangci-lint run`.

### CODE REVIEW

File: internal/federation/authenticate.go [hunk 1]
```diff
@@ -45,11 +48,45 @@ var (
 	}
 )
 
+type PubKeyResponse struct {
```
Comment (NyaaaWhatsUpDoc) [inline internal/federation/authenticate.go:51]: Quite a subjective note here; I think ts name sort of indicates this is the response we'd get back from an endpoint which is a little confusing from the AuthenticateFederatedRequest() function.

What do you think of changing the name to something along the lines of `AuthOwner`, and remove the `PubKeyExpired` boolean in favour of an account receiver function along the lines of `func (a *Account) PubKeyExpired()`. Just helps to simplify the code here a little and reduce doubled-up data

File: internal/federation/authenticate.go [hunk 2]
```diff
@@ -196,42 +265,93 @@ func (f *federator) deref(
...
-	if errWithCode == nil {
-		l.Trace("public key cached, no dereference needed")
-		return reqAcctURI, pubKey, nil
+	pubKeyResponse, errWithCode := f.derefPubKeyDBOnly(ctx, pubKeyIDStr)
+	if errWithCode != nil {
+		return nil, errWithCode
 	}
 
-	l.Trace("public key not cached, trying dereference")
+	if pubKeyResponse != nil && !pubKeyResponse.PubKeyExpired {
```
Comment (NyaaaWhatsUpDoc) [inline internal/federation/authenticate.go:273]: Following the if else flow along here is a little confusing, what do you think of a switch instead along the lines of:
```

expired := (pubKeyRsp != nil && pubKeyRsp.Expired)

switch {
// Yet-to-be cached public key
case pubKeyRsp == nil:
    log.Trace(...)

// Existing account with useable public key
case !expired:
    log.Trace(...)
    return pubKeyRsp, nil

// Existing account with an expired public key
case expired:
    log.Infof(...)
}

// ...

pubKey, pubKeyOwner, err := parsePubKeyBytes(...)
if err != nil {
    return nil, err
}

if !expired {
    // This is a new (to us) public key and owner URI
    return PublicKeyResponse{...}, nil
}

```

Obviously with var names changed a little if you make a change with the above struct ^

File: internal/federation/authenticate.go (hunk 1 above)
Comment (tsmethurst) [reply internal/federation/authenticate.go]: Good idea! I fixed 'er up :)

File: internal/federation/authenticate.go (hunk 2 above)
Comment (tsmethurst) [reply internal/federation/authenticate.go]: Honestly I wrote this logic like 4 different ways originally and wasn't super happy with any of them :') But the switch statement does seem neater yeah, so I fixed it up :)

**Comment by NyaaaWhatsUpDoc** [comment]
nice work :cowboy_hat_face:

**Pull request merged**